Bitwise builtin

mmsc2 opened the pull request.

Commit:
WIP: Implementing range_check builtin
- Adding constant declarations
- Adding memory error
- Implementing first draft of builtin

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo_vm.go into range-check-builtin

Commit:
Implementing range_check builtin
- Add Bits function on lambdaworks API
- Implements builtin_runner interface on range_check builtin.

Commit:
Testing range_check builtin
- Renamed to Uppercase some attributes

Commit:
Fixing felt inicialization

Commit:
Fixing felt inicialization

toni-calvin committed:
Update pkg/runners/builtin_runner/range_check.go

Co-authored-by: fmoletta <99273364+fmoletta@users.noreply.github.com>

Commit:
Creating class method One and Zero

Commit:
Deleting not used fields in RangeCheckBuiltinRunner

toni-calvin committed:
Update pkg/runners/builtin_runner/range_check.go

Co-authored-by: fmoletta <99273364+fmoletta@users.noreply.github.com>

toni-calvin committed:
Update pkg/runners/builtin_runner/range_check.go

Co-authored-by: fmoletta <99273364+fmoletta@users.noreply.github.com>

Commit:
Defininf builtin.base as Relocatable

Commit:
Merge branch 'range-check-builtin' of github.com:lambdaclass/cairo_vm.go into range-check-builtin

Commit:
Using lambdaworks shl method

Commit:
Deleting unnecessary method (for now)

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo_vm.go into range-check-builtin

Commit:
Improve error handeling

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo_vm.go into range-check-builtin

Commit:
Added check range documentation

toni-calvin committed:
Update pkg/lambdaworks/lambdaworks_test.go

Co-authored-by: Gabriel Bosio <38794644+gabrielbosio@users.noreply.github.com>

toni-calvin committed:
Update pkg/lambdaworks/lambdaworks_test.go

Co-authored-by: Gabriel Bosio <38794644+gabrielbosio@users.noreply.github.com>

Commit:
Merge branches 'range-check-builtin' and 'main' of github.com:lambdaclass/cairo_vm.go into range-check-builtin

Commit:
Typos

Commit:
Fixing some merge conflicts

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo_vm.go into range-check-builtin

Commit:
Fixing TestValidateExistingMemoryForRangeCheckOutsideBounds

Commit:
Remove unused lambdaworks methods

mmsc2 committed:
bitwise builtin

mmsc2 committed:
Bitwise builtin

mmsc2 committed:
operation bitwise

mmsc2 committed:
Expose and or bitwise operations

mmsc2 committed:
Merge main

mmsc2 committed:
Export pow or and functions

mmsc2 committed:
Deduce memory cell

mmsc2 committed:
Add test

mmsc2 committed:
test for deduce memory

mmsc2 committed:
Add more tests, delete debug info

mmsc2 committed:
Delete git merge message from readme

mmsc2 committed:
Add tests for xor

mmsc2 committed:
Add new line in lambdaworks.h

mmsc2 committed:
Merge branch 'main' into BitwiseBuiltin

mmsc2 committed:
debug test

mmsc2 committed:
Address comments

mmsc2 committed:
Add builtin to cairo un

mmsc2 committed:
Add builtin to cairo un

mmsc2 committed:
Add integration test

mmsc2 committed:
Delete range check stuff

mmsc2 committed:
Delete range check stuff

mmsc2 committed:
Delete range check stuff from readme

mmsc2 committed:
Delete log from test

mmsc2 committed:
FIx errors

mmsc2 committed:
Fix tests

mmsc2 committed:
remove todo

mmsc2 committed:
Documentation

mmsc2 committed:
Merge branch 'main' into BitwiseBuiltin

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,106 @@
+type BitwiseBuiltinRunner struct {
+	base             memory.Relocatable
+	included         bool
+	CellsPerInstance uint
```
We can remove this field and use an constant instead

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+func (b *BitwiseBuiltinRunner) Name() string {
+	return "bitwise_builtin"
```
```suggestion
	return "bitwise"
```

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,106 @@
+	included         bool
+	CellsPerInstance uint
+	Ratio            *uint
+	BitwiseBuiltin   BitwiseInstanceDef
```
We don't need this

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,106 @@
+
+type BitwiseInstanceDef struct {
+	Ratio      *uint
+	TotalNBits uint
```
TotalNBits is always 251 so we can use a constant instead

mmsc2 committed:
Documentation

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,106 @@
+
+func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, segments *memory.Memory) (*memory.MaybeRelocatable, error) {
+	index := address.Offset % b.CellsPerInstance
+	if index < 1 {
```
This should be <= 1. 
Even better, we could set the constant `INPUT_CELLS_PER_INSTANCE = 2` and use it in this if

mmsc2 replied:
INPUT_CELLS_PER_INSTANCE shouldn't be 1 ?

mmsc2 replied:
or you mean setting in two, because it contemplates the case <= 1 ?

fmoletta replied:
Yes, the number of  input cells for bitwise is 2 so adding the constant and using it here will make it easier to document

fmoletta commented on pkg/vm/memory/memory.go:
```diff
@@ -24,45 +24,43 @@ type ValidationRule func(*Memory, Relocatable) ([]Relocatable, error)
+	Data              map[Relocatable]MaybeRelocatable
+	NumSegments       uint
+	ValidationRules   map[uint]ValidationRule
+	ValidatedAdresses AddressSet
```
Dont do this

mmsc2 replied:
this is from the PR of range check i guess, i had to merge that branch into this one as i needed somethings

mmsc2 committed:
Delte debug info

fmoletta reviewed (commented):
Looks good, suggested some minor changes but it should be good to merge after adressing them and adding some integration tests
We should add some integration tests that use CairoRun to run a program that uses bitwise operation, you can get some cairo programs from the cairo-vm repo
This PR contains changes from the range_check PR, pease remove them so we can merge it, otherwise we have to wait till range_check is finished and merged

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,99 @@
+
+func NewBitwiseBuiltinRunner(included bool) BitwiseBuiltinRunner {
+	return BitwiseBuiltinRunner{
+		base:     memory.NewRelocatable(0, 0),
```
nit: We can omit this field when creating the struct and it will take the zero-value (which is 0,0)

mmsc2 committed:
Fix Readme

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,98 @@
+	num_y_felt, y_is_felt := num_y.GetFelt()
+
+	if x_is_felt && y_is_felt {
+		// TODO: check if this conversion if valid
```
Remove TODO

mmsc2 committed:
Address coments

fmoletta commented on README.md:
```diff
@@ -1666,6 +1666,106 @@ With all of our builtin logic integrated into the codebase, we can implement any
+
+The method `DeducedMemoryCell` fetches from memory the operands and  performes the operations. 
+- If the index is lesser than two then the methods just return nil
+- After the operands are fetch it is checked that both are felts because bitwise can not be perform on relocatables or nil values.
```
```suggestion
- After the operands are fetched it checks that both of them are felts because bitwise can not be performed on relocatable or nil values.
```

fmoletta commented on README.md:
````diff
@@ -1666,6 +1666,106 @@ With all of our builtin logic integrated into the codebase, we can implement any
+}
+```
+
+The method `DeducedMemoryCell` fetches from memory the operands and  performes the operations. 
````
```suggestion
The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations.:
```

fmoletta commented on README.md:
````diff
@@ -1666,6 +1666,106 @@ With all of our builtin logic integrated into the codebase, we can implement any
+```
+
+The method `DeducedMemoryCell` fetches from memory the operands and  performes the operations. 
+- If the index is lesser than two then the methods just return nil
````
```suggestion
- If the index is less than the number of input cells then the methods just return nil
```

fmoletta commented on README.md:
````diff
@@ -1666,6 +1666,106 @@ With all of our builtin logic integrated into the codebase, we can implement any
+- The index is used to know which bitwise operation is going to be performed, if is 2 then `and` is executed, if is 3 then `xor` and if is a 4 then `or` is executed
+- Otherwise nil value is returned
+
+```go 
````
Add the definition of the constants to this code block

fmoletta commented on README.md:
```diff
@@ -1666,6 +1666,106 @@ With all of our builtin logic integrated into the codebase, we can implement any
+#### Bitwise builtin
+
+This builtin provides a way to work with the basic bit operations `and`, `or` and `xor`. It implements the basic builtin interface methods:
+
```
Add BitWiseBuiltinRunner type definton

fmoletta commented on README.md:
```diff
@@ -1666,6 +1666,118 @@ With all of our builtin logic integrated into the codebase, we can implement any
-#### Hints
 
-TODO
+### Implementing each builtin runner 
```
```suggestion
#### Implementing each builtin runner 
```
This has the same hierarchy as Write your own Vm title

fmoletta commented on README.md:
```diff
@@ -1666,6 +1666,118 @@ With all of our builtin logic integrated into the codebase, we can implement any
+### Implementing each builtin runner 
+
+
+#### Bitwise builtin
```
```suggestion
##### Bitwise builtin
```

fmoletta commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,95 @@
+
+const CELLS_PER_INSTANCE = 5
+const TOTAL_N_BITS = 251
+const INPUT_CELLS_PER_INSTANCE = 2
```
```suggestion
const BITWISE_CELLS_PER_INSTANCE = 5
const BITWISE_TOTAL_N_BITS = 251
const BIWISE_INPUT_CELLS_PER_INSTANCE = 2
```
As all the builtins are in the same package, constant's names can clash

gabrielbosio commented on README.md:
````diff
@@ -1666,6 +1666,118 @@ With all of our builtin logic integrated into the codebase, we can implement any
+}
+```
+
+For the `InitializeSegments` we just add a segment to the memory and store in the base attribute, the first adress of the segment. 
````
```suggestion
For the `InitializeSegments` we just add a segment to the memory and store in the base attribute, which is the first address of the segment. 
```

toni-calvin replied:
I think it would be nicer something like: 

```suggestion
In the `InitializeSegments` method, we add a new segment to the memory and set the base attribute as the first segment's address. 
```

gabrielbosio replied:
Agree

gabrielbosio commented on README.md:
````diff
@@ -1666,6 +1666,118 @@ With all of our builtin logic integrated into the codebase, we can implement any
+}
+```
+
+The method `DeducedMemoryCell` fetches the operands from memory performes the following operations: 
````
```suggestion
The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations: 
```

gabrielbosio commented on README.md:
````diff
@@ -1666,6 +1666,118 @@ With all of our builtin logic integrated into the codebase, we can implement any
+```
+
+The method `DeducedMemoryCell` fetches the operands from memory performes the following operations: 
+- If the index is less than the number of input cells then the methods just return nil
````
```suggestion
- If the index is less than the number of input cells then the method returns nil
```

gabrielbosio commented on README.md:
```diff
@@ -1666,6 +1666,118 @@ With all of our builtin logic integrated into the codebase, we can implement any
+The method `DeducedMemoryCell` fetches the operands from memory performes the following operations: 
+- If the index is less than the number of input cells then the methods just return nil
+- After the operands are fetched it is checks that both of them are felts because bitwise can not be performed on relocatable or nil values.
+- If the number of bits of any operands is greater than TOTAL_N_BITS the method fails because we are out of the field.
```
```suggestion
- If the number of bits of any operand is greater than `TOTAL_N_BITS` the method fails because we are out of the field.
```

gabrielbosio commented on pkg/lambdaworks/lambdaworks.go:
```diff
@@ -159,3 +161,44 @@ func (a Felt) Div(b Felt) Felt {
+	var b_c C.felt_t = b.toC()
+	C.felt_pow(&a_c[0], &b_c[0], &result[0])
+	return fromC(result)
+}
```
Why is this function needed for this builtin?

mmsc2 replied:
is not, there was a confusion with the ^ operand that i though that it was exponentiation but it was xor. i left it because maybe will be used in the future

gabrielbosio replied:
If this function is needed, please move this and its tests to a new PR

toni-calvin replied:
Done here: #141

gabrielbosio commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,95 @@
+		return nil, nil
+	}
+
+	x_addr := memory.NewRelocatable(address.SegmentIndex, address.Offset-index)
```
```suggestion
	x_addr := address.SubUint(index)
```

toni-calvin replied:
Done!

gabrielbosio commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,95 @@
+	}
+	num_x, err := segments.Get(x_addr)
+	if err != nil {
+
```
Please, remove this newline

gabrielbosio commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,95 @@
+
+	num_y, err := segments.Get(y_addr)
+	if err != nil {
+
```
Please, remove this newline

gabrielbosio commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,95 @@
+
+	if x_is_felt && y_is_felt {
+		if num_x_felt.Bits() > BITWISE_TOTAL_N_BITS {
+			return nil, errors.New("Expected Intenger x to be smaller than 2^(total_n_bits)")
```
```suggestion
			return nil, fmt.Errorf("Expected Intenger x to be smaller than 2^%d", BITWISE_TOTAL_N_BITS)
```

toni-calvin replied:
Done!

gabrielbosio commented on pkg/runners/builtin_runner/bitwise.go:
```diff
@@ -0,0 +1,95 @@
+			return nil, errors.New("Expected Intenger x to be smaller than 2^(total_n_bits)")
+		}
+		if num_y_felt.Bits() > BITWISE_TOTAL_N_BITS {
+			return nil, errors.New("Expected Intenger y to be smaller than 2^(total_n_bits)")
```
```suggestion
			return nil, errors.New("Expected Intenger y to be smaller than 2^%d", BITWISE_TOTAL_N_BITS)
```

toni-calvin replied:
Done!

gabrielbosio commented on pkg/runners/builtin_runner/bitwise_test.go:
```diff
@@ -0,0 +1,172 @@
+
+	mem.Memory.Insert(rel1, m1)
+	mem.Memory.Insert(rel2, m2)
+	mem.Memory.Insert(rel3, m3)
```
I think we can remove this insertion because we would be inserting a zero

gabrielbosio commented on pkg/runners/builtin_runner/bitwise_test.go:
```diff
@@ -0,0 +1,172 @@
+
+	mem.Memory.Insert(rel1, m1)
+	mem.Memory.Insert(rel2, m2)
+	mem.Memory.Insert(rel3, m3)
```
I think we can remove this insertion because we would be inserting a zero

gabrielbosio commented on pkg/runners/builtin_runner/bitwise_test.go:
```diff
@@ -0,0 +1,172 @@
+
+	mem.Memory.Insert(rel1, m1)
+	mem.Memory.Insert(rel2, m2)
+	mem.Memory.Insert(rel3, m3)
```
I think we can remove this insertion because we would be inserting a zero

gabrielbosio commented on pkg/runners/builtin_runner/bitwise_test.go:
```diff
@@ -0,0 +1,172 @@
+
+	mem.Memory.Insert(rel1, m1)
+	mem.Memory.Insert(rel2, m2)
+	mem.Memory.Insert(rel3, m3)
```
I think we can remove this insertion because we would be inserting a zero

toni-calvin replied:
True, done, here and in the rest of the builtin code

gabrielbosio commented on pkg/runners/cairo_runner.go:
```diff
@@ -3,6 +3,7 @@ package runners
 import (
 	"errors"
 
+	builtinrunner "github.com/lambdaclass/cairo-vm.go/pkg/runners/builtin_runner"
```
For clarity, we can move the builtin to a new `builtins` folder

toni-calvin replied:
Done!

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo-vm.go into BitwiseBuiltin

Commit:
Move builtin files inside folder

Commit:
Return builtin documentation

Commit:
Return builtin documentation

toni-calvin committed:
Apply suggestions from code review

Co-authored-by: Gabriel Bosio <38794644+gabrielbosio@users.noreply.github.com>

Commit:
Review PR
* Remove unused method
* Improve testing in new lambdaworks lib methods

Commit:
Added constant with builtin name

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo-vm.go into BitwiseBuiltin

Commit:
Use of felt add function in deduce memory cells test + simplify the test

Commit:
Remove unnecessary newlines + added some error control

Commit:
Fix problem with big (and negative) identifier value.
This commit improves error handling
This commit deserialize identifiers making now identifier.value as lambdaworks.Felt

Commit:
Added error improvements and deleted insertions of default values (felt.0) in memory

toni-calvin commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+
+	if x_is_felt && y_is_felt {
+		if num_x_felt.Bits() > BITWISE_TOTAL_N_BITS {
+			return nil, ErrFeltBiggerThanPowerOfTwo(num_x_felt, BITWISE_TOTAL_N_BITS)
```
Error handling example

gabrielbosio commented on pkg/builtins/bitwise_test.go:
```diff
@@ -0,0 +1,146 @@
+
+	mem.Memory.Insert(rel1, m1)
+	mem.Memory.Insert(rel2, m2)
+	mem.Memory.Insert(rel3, m3)
```
I think this line can be removed, because `Memory.Get` returns 0 by default

fmoletta replied:
It doesn't return 0 by default, it returns a zero value when it fails, if we don't insert the 0, Memory.Get will still fail (aka return (0, false))

toni-calvin replied:
In fact it fails, we are returning this: 

`
value, ok := m.data[addr]
if !ok {
 return nil, errors.New("Memory Get: Value not found")
}
`

But in the tests that position was not being loaded from memory so it did not fail, I add again the values

gabrielbosio commented on pkg/builtins/bitwise_test.go:
```diff
@@ -0,0 +1,146 @@
+	m2 := memory.NewMaybeRelocatableFelt(lambdaworks.FeltFromUint64(0))
+
+	mem.Memory.Insert(rel1, m1)
+	mem.Memory.Insert(rel2, m2)
```
I think this line can be removed, because `Memory.Get` returns 0 by default

fmoletta replied:
Memory.Get would still fail

gabrielbosio reviewed (approved):
Approved with a few comments

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+
+func ErrFeltBiggerThanPowerOfTwo(felt lambdaworks.Felt, exponent uint) error {
+	upperBound := uint64(math.Pow(2, float64(exponent)))
+	return BitwiseError(errors.Errorf("Expected felt %d to be smaller than %d", felt, upperBound))
```
```suggestion
	return BitwiseError(errors.Errorf("Expected felt %d to be smaller than  2**%d", felt, upperBound))
```
This looks clearer to the user imo. Also if the value is constant we can hardcode it in the error message

toni-calvin replied:
100% agree, I though of doing a more abstract composition if there were other callers, but I agree is clearer

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+	x_addr, err := address.SubUint(index)
+	if err != nil {
+		return nil, BitwiseError(err)
+	}
```
Here index will never be bigger than addr.Offset so we can ignore the error (with a comment)

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+	}
+	num_x, err := segments.Get(x_addr)
+	if err != nil {
+		return nil, BitwiseError(err)
```
```suggestion
		return nil, nil
```
Correct behaviour

toni-calvin replied:
Why error should not be propagated?

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+	num_y, err := segments.Get(y_addr)
+	if err != nil {
+		return nil, BitwiseError(err)
+	}
```
```suggestion
	y_addr := x_addr.AddUint(1)
	num_y, err := segments.Get(y_addr)
	if err != nil {
		return nil, nil
	}
```
Correct behaviour

fmoletta commented on pkg/runners/cairo_runner.go:
```diff
@@ -4,6 +4,7 @@ import (
 	"errors"
 
 	"github.com/lambdaclass/cairo-vm.go/pkg/builtins"
+	builtinrunner "github.com/lambdaclass/cairo-vm.go/pkg/builtins"
```
why

fmoletta commented on pkg/vm/program.go:
```diff
@@ -21,7 +32,19 @@ func DeserializeProgramJson(compiledProgram parser.CompiledJson) Program {
 	}
 	program.Builtins = compiledProgram.Builtins
-	program.Identifiers = &compiledProgram.Identifiers
+	program.Identifiers = make(map[string]Identifier)
+	for key, identifier := range compiledProgram.Identifiers {
+		var programIdentifier Identifier
+		programIdentifier.FullName = identifier.FullName
+		programIdentifier.Members = identifier.Members
+		programIdentifier.Size = identifier.Size
+		programIdentifier.Decorators = identifier.Decorators
+		programIdentifier.PC = identifier.PC
+		programIdentifier.Type = identifier.Type
+		programIdentifier.CairoType = identifier.CairoType
+		programIdentifier.Value = lambdaworks.FeltFromDecString(identifier.Value.String())
+		program.Identifiers[key] = programIdentifier
+	}
```
Could we use the same Identifier struct for the parser and the program?

toni-calvin replied:
We use `err = json.Unmarshal(byteValue, &cJson)` to unmarshall json into parser struct, but can't define a type felt there. It is possible but we would have to iterate through field to unmarshall them. This way we can point to reference while the structure is the same or add conversion functions when needed.

Commit:
Improve error

Commit:
Message format

Commit:
PR review

Commit:
PR review

Commit:
Merge branch 'BitwiseBuiltin' of github.com:lambdaclass/cairo-vm.go into BitwiseBuiltin

toni-calvin committed:
Apply suggestions from code review

Co-authored-by: fmoletta <99273364+fmoletta@users.noreply.github.com>

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	num_x_felt, x_is_felt := num_x.GetFelt()
+	num_y_felt, y_is_felt := num_y.GetFelt()
```
We can use `memory.GetFelt` to avoid having to use memory.Get + num.GetFelt

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+	}
+}
+
+func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, segments *memory.Memory) (*memory.MaybeRelocatable, error) {
```
```suggestion
func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, memory *memory.Memory) (*memory.MaybeRelocatable, error) {
```
Wrong name, `mem` could also be used if it conflicts with the package name

fmoletta commented on pkg/builtins/bitwise.go:
```diff
@@ -0,0 +1,104 @@
+	// Error can't happen, index is never bigger than address offset
+	// if err != nil {
+	// 	return nil, BitwiseError(err)
+	// }
```
Remove

fmoletta commented on README.md:
```diff
@@ -2404,7 +2404,115 @@ TODO
+const TOTAL_N_BITS = 251
+const INPUT_CELLS_PER_INSTANCE = 2
+
+func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, segments *memory.Memory) (*memory.MaybeRelocatable, error) {
```
Please update this

fmoletta commented on README.md:
```diff
@@ -2416,4 +2524,7 @@ TODO
+#### Implementing each builtin runner 
+
+
+
```
Remove

Commit:
Merge branch 'main' of github.com:lambdaclass/cairo-vm.go into BitwiseBuiltin

Commit:
Resolve comments on bitwise builtin
